kernel/src/actions/visitors.rs

```diff
@@ -724,4 +797,116 @@ mod tests {
+            r#"{"domainMetadata":{"domain": "zach4","configuration":"old_cfg4","removed": true}}"#,
+            r#"{"domainMetadata":{"domain": "zach7","configuration":"cfg7","removed": true}}"#,
+            r#"{"domainMetadata":{"domain": "zach8","configuration":"cfg8","removed": false}}"#,
+        ]
+        .into();
+        let batch2 = parse_json_batch(json_strings);
+        let mut domain_metadata_visitor = DomainMetadataVisitor::default();
+        // visit batch 1 then 2 (would assume batch 1 is 'newer' in the log)
+        domain_metadata_visitor
+            .visit_rows_of(batch1.as_ref())
+            .unwrap();
+        domain_metadata_visitor
+            .visit_rows_of(batch2.as_ref())
+            .unwrap();
+        let actual = domain_metadata_visitor.domain_metadatas;
+        let expected = DomainMetadataSet::from([
+            DomainMetadata {
+                domain: "zach1".to_string(),
+                configuration: "cfg1".to_string(),
+                removed: true,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach2".to_string(),
+                configuration: "cfg2".to_string(),
+                removed: false,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach3".to_string(),
+                configuration: "cfg3".to_string(),
+                removed: true,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach4".to_string(),
+                configuration: "cfg4".to_string(),
+                removed: false,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach5".to_string(),
+                configuration: "cfg5".to_string(),
+                removed: true,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach6".to_string(),
+                configuration: "cfg6".to_string(),
+                removed: false,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach7".to_string(),
+                configuration: "cfg7".to_string(),
+                removed: true,
+            }
+            .into(),
+            DomainMetadata {
+                domain: "zach8".to_string(),
+                configuration: "cfg8".to_string(),
+                removed: false,
+            }
+            .into(),
+        ]);
+        assert_eq!(actual, expected);
+
+        // test filtering
+        let mut domain_metadata_visitor = DomainMetadataVisitor::new(Some("zach3".to_string()));
+        domain_metadata_visitor
+            .visit_rows_of(batch1.as_ref())
+            .unwrap();
+        domain_metadata_visitor
+            .visit_rows_of(batch2.as_ref())
+            .unwrap();
+        let actual = domain_metadata_visitor.domain_metadatas;
+        let expected = DomainMetadataSet::from([DomainMetadata {
```

Comment:
The visitor expects things in reverse order, so `batch1` is "newer" than `batch2` in this case (i think, maybe rename them to make that clear).

So then shouldn't the data in `batch1` take priority, which would imply that at the end we should get: `{"domainMetadata":{"domain": "zach3","configuration":"cfg3","removed": true}}`

Reply:
yep left over from my incorrect remove handling (and i did a rename just to `commit_1` and `commit_0`)